### PR TITLE
improvement: show workspace completions for short queries

### PR DESCRIPTION
**Pull request by kasiaMarek**
resolves: https://github.com/scalameta/metals/issues/5619

Previously: for query shorter than 3, we'd only show exact matches in completions (for classpath/workspace symbols), now we show all that have the correct prefix. Not exact matches from classpath are capped at 10, additionally we cap workspace symbols for those short queries at 100. If no prefix matches are found, we retry allowing for full fuzzy match.

Also fixes sorting in Scala 3, so keyword completions are offered before workspace/classpath ones.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/WorkspaceSymbolProvider.scala [hunk 1]
```diff
@@ -203,12 +205,16 @@ final class WorkspaceSymbolProvider(
...
       if query.matches(symbol.symbol)
     } {
-      visitor.visitWorkspaceSymbol(
-        path,
-        symbol.symbol,
-        symbol.kind,
-        symbol.range,
-      )
+      count += 1
+      if (query.isExact && count >= MaxWorkspaceMatchesForShortQuery) return
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/WorkspaceSymbolProvider.scala:209]: Not super fond of the return. The alternative is to use a recursion after the for comprehension that would return an iterator.

Something like:
```scala
val symbols: Iterator[Symbol] = for {...}

@tailrec
def loop(currentNum: Int): Unit = {
  if ( ! (query.isExact ||  count >= MaxWorkspaceMatchesForShortQuery) && symbols.hasNext) 
    val sym = symbols.next()
    visitor.visitWorkspaceSymbol(
          path,
          symbol.symbol,
          symbol.kind,
          symbol.range,
        )
     loop(currentNum + 1)
   }
}
```
might not be exactly right.

File: mtags-shared/src/main/scala/scala/meta/internal/metals/ClasspathSearch.scala [hunk 2]
```diff
@@ -57,21 +57,27 @@ class ClasspathSearch(
...
+    val result =
+      for {
+        pkg <- packages.iterator
+        if pkg.packages.exists(shouldVisitPackage)
+        if !isCancelled()
+        if query.matches(pkg.bloom)
+        classfile <- pkg.members
+        if classfile.isClassfile
+        isMatch = {
+          if (query.isExact && !isRetry)
```
Comment (tgodzik) [inline mtags-shared/src/main/scala/scala/meta/internal/metals/ClasspathSearch.scala:72]: This seems to drop fuzzy match entirely.

Should we have a `Prefix` search type instead of reusing Exact? We could do the prefix if we have query.length < 3 ? So that the current behaviour stays but we show things better if we have 1 or 2 characters?

File: mtags-shared/src/main/scala/scala/meta/internal/metals/Fuzzy.scala [hunk 3]
```diff
@@ -124,16 +124,32 @@ class Fuzzy {
...
+      }
+      true
     }
   }
 
+  def findStartIndex(symbol: CharSequence, lastIndex: Int): Int = {
+    var start = 0
+    var i = 0
+    while (i < lastIndex) {
+      if (symbol.charAt(i) == '$') {
```
Comment (tgodzik) [inline mtags-shared/src/main/scala/scala/meta/internal/metals/Fuzzy.scala:145]: Why is this needed? Could you add a comment?

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala [hunk 4]
```diff
@@ -37,6 +39,11 @@ class CompletionKeywordSuite extends BaseCompletionSuite {
       |}
       |""".stripMargin,
     "",
+    compat = Map(
+      "3" -> """|transient scala (commit: '')
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:43]: That's inside a comment, so we should not show anything here :thinking:

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala [hunk 5]
```diff
@@ -56,6 +63,11 @@ class CompletionKeywordSuite extends BaseCompletionSuite {
       |}
       |""".stripMargin,
     "",
+    compat = Map(
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:66]: Same here

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala [hunk 6]
```diff
@@ -183,7 +199,8 @@ class CompletionKeywordSuite extends BaseCompletionSuite {
       |""".stripMargin,
     """|val
        |var
-       |""".stripMargin
+       |""".stripMargin,
+    topLines = Some(2)
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:203]: Let's not add toplines in this file maybe so that we see exactly what changed? Or are there so many additional files now?

File: tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala [hunk 7]
```diff
@@ -60,7 +60,25 @@ class CompletionPatternSuite extends BaseCompletionSuite {
       |    case ma@@
       |  }
       |}""".stripMargin,
-    ""
+    "",
+    compat = Map(
+      scala3PresentationCompilerVersion ->
+        """|main scala
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: So this would show before if we wrote for example `mai` ? I think we shouldn't show anything in this case, but might be a different bug, not to fix here.

File: tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala [hunk 8]
```diff
@@ -103,7 +103,8 @@ class CompletionSnippetSuite extends BaseCompletionSuite {
       |}
       |""".stripMargin,
     """|MyType
-       |""".stripMargin
+       |""".stripMargin,
+    topLines = Some(1)
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala:107]: If there is too much results, let's at least show like 5

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala [hunk 9]
```diff
@@ -429,14 +429,15 @@ class CompletionSuite extends BaseCompletionSuite {
...
-    "import-star-multi-import",
+    "import-star-multi-import".tag(IgnoreForScala3CompilerPC),
     """
       |import scala.collection.immutable.List.{range => r, *@@}
       |""".stripMargin,
     "",
     compat = Map(
       "3" ->
         """|*
+           |*: scala
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:440]: This seems incorrect, error is reported in this case

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala [hunk 10]
```diff
@@ -1723,7 +1725,8 @@ class CompletionSuite extends BaseCompletionSuite {
     "TTT[A] = O.TTT",
     compat = Map(
       "3" -> "TTT[A <: Int] = List[A]"
-    )
+    ),
+    topLines = Some(1)
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:1729]: Let's add some more lines to better see what changed.

**Review by tgodzik — CHANGES_REQUESTED** [review]
Thanks for the great work! I left some comments to make sure the change is what we want.

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 4 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:43]: Agreed but this was broken all along. Just the test did not test it properly. Let's maybe do it as a followup.

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 6 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:203]: It is quite a lot of completions and those differ between scala versions.

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 4 above)
Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:43]: Ach, why did the test not show it before? Just because there were fewer than 3 letters?

File: tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala (hunk 7 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: yes

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 4 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:43]: Yes, it only searched for exact matches in workspace/classpath when less than 3 chars.

File: mtags-shared/src/main/scala/scala/meta/internal/metals/ClasspathSearch.scala (hunk 2 above)
Comment (kasiaMarek) [reply mtags-shared/src/main/scala/scala/meta/internal/metals/ClasspathSearch.scala:72]: I don't drop fuzzy match altogether for workspace symbols still fuzzy match is used, only for classpath I used prefix. Also it fallback to fuzzy if using prefix nothing was found (since we don't want empty completions). But I can use the same logic for both classpath and workspace of using prefix for < 3 chars and using fuzzy as fallback.

Comment (tgodzik) [reply mtags-shared/src/main/scala/scala/meta/internal/metals/ClasspathSearch.scala:72]: I think overall fuzzy makes more sense since it's easy to forget exact name of a class. But with 2 or 1 character it's useless, so I would do prefix only then.

Of course we could gather some opinions on discord or scala contributors.

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 4 above)
Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:43]: ok sure, let's create an issue then

File: tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala (hunk 7 above)
Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: would be good to fix in a separate issue, though it seems like something that we dealt before with? @jkciesluk ?

Comment (jkciesluk) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: It's shown here because `Completions.advancedCompletions` for patterns in `match-case` are not marked as `exclusive` (if they were, we wouldn’t suggest eg. objects with unapply methods).

Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: Could we make sure only things with unapply show up?

Comment (jkciesluk) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: Yeah, some time ago I was trying to implement completions for them that would match the result type of the `unapply` method, but didn’t manage to finish it. Adding just completion for object where `unapply` is defined (with maybe not matching parameter type) should be a lot easier.

Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: That would already be better I think, but anyway, not related

Comment (jkciesluk) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: Maybe we can leave this `compat` here now and remove after adding this feature?

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 4 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:43]: https://github.com/scalameta/metals/issues/5971

File: tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala (hunk 7 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionPatternSuite.scala:66]: https://github.com/scalameta/metals/issues/5972

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala (hunk 9 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:440]: After I changed the logic no error is reported anymore. We should not do workspace/classpath search in this case at all, but again it's a separate issue.

File: mtags-shared/src/main/scala/scala/meta/internal/metals/Fuzzy.scala [hunk 11]
```diff
@@ -124,13 +137,30 @@ class Fuzzy {
...
-  def isExactMatch(query: String, filename: CharSequence): Boolean = {
-    val sb = lastIndex(filename)
-    val sa = sb - query.length()
-    if (sa < 0) {
-      false
-    } else {
-      exactMatch(query, filename, sa, sb)
+  private def matchesNamePrefix(
+      query: CharSequence,
+      qa: Int,
```
Comment (tgodzik) [inline mtags-shared/src/main/scala/scala/meta/internal/metals/Fuzzy.scala:142]: Might be good to add better names for these now.

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala [hunk 12]
```diff
@@ -198,11 +222,16 @@ class CompletionKeywordSuite extends BaseCompletionSuite {
       |}
       |""".stripMargin,
     """|given (commit: '')
+       |given_FromString_Int - scala.util.CommandLineParser.FromString (commit: '')
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:225]: How did these get here? :thinking: Those are member and we normally don't show those. Should we create a follow up?

File: tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala [hunk 13]
```diff
@@ -103,7 +103,12 @@ class CompletionSnippetSuite extends BaseCompletionSuite {
       |}
       |""".stripMargin,
     """|MyType
-       |""".stripMargin
+       |""".stripMargin,
+    compat = Map(
+      "2.11" -> """|MyType
+                   |MTOM
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala:109]: Any idea, where is this coming from?

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala (hunk 9 above)
Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:440]: Could we create a follow up?

**Review by tgodzik — APPROVED** [review]
Looks good, just some minor comments, otherwise we can emrge

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala (hunk 9 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:440]: https://github.com/scalameta/metals/issues/5991

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 12 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:225]: https://github.com/scalameta/metals/issues/5992

File: tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala (hunk 13 above)
Comment (kasiaMarek) [reply tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala:109]: Classpath:
```
javax/xml/ws/soap/MTOM.class
javax/xml/ws/soap/MTOMFeature.class
```
Seems to be a Java 8 thing.

Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionSnippetSuite.scala:109]: Thanks for checking!
